File: cli/cmd/kill.go
package cmd

import (
	"fmt"
	"strconv"

	"github.com/productdevbook/port-killer/cli/internal/scanner"
	"github.com/spf13/cobra"
)

var forceKill bool

var killCmd = &cobra.Command{
	Use:   "kill <port>",
	Short: "Kill process listening on a port",
	Long:  `Kill the process that is listening on the specified port. Uses SIGTERM by default, SIGKILL with --force.`,
	Args:  cobra.ExactArgs(1),
	RunE:  runKill,
}

func init() {
	killCmd.Flags().BoolVarP(&forceKill, "force", "f", false, "Force kill with SIGKILL")
}

func runKill(cmd *cobra.Command, args []string) error {
	port, err := strconv.Atoi(args[0])
	if err != nil {
		return fmt.Errorf("invalid port number: %s", args[0])
	}

	s := scanner.New()
	ports, err := s.Scan()
	if err != nil {
		return fmt.Errorf("failed to scan ports: %w", err)
	}

	// Find the process on this port
	var target *scanner.Port
	for _, p := range ports {
		if p.Port == port {
			target = &p
			break
		}
	}

	if target == nil {
		return fmt.Errorf("no process found listening on port %d", port)
	}

	// Kill the process
	if err := s.Kill(target.PID, forceKill); err != nil {
		return fmt.Errorf("failed to kill process %d: %w", target.PID, err)
	}

	action := "Killed"
	if forceKill {
		action = "Force killed"
	}
	fmt.Printf("%s %s (PID %d) on port %d\n", action, target.Process, target.PID, target.Port)

	return nil
}

File: backend/cli/src/commands/kill.rs
//! Kill command - terminate process on a port.

use anyhow::{bail, Result};
use portkiller_core::{PortScanner, ProcessKiller};

pub async fn run(port: u16, force: bool) -> Result<()> {
    let scanner = PortScanner::new();
    let ports = scanner.scan().await?;

    // Find the port
    let port_info = ports.iter().find(|p| p.port == port);

    let port_info = match port_info {
        Some(p) => p,
        None => {
            println!("No process found on port {}.", port);
            return Ok(());
        }
    };

    let pid = port_info.pid;
    let process_name = &port_info.process_name;

    println!(
        "Killing {} (PID: {}) on port {}{}...",
        process_name,
        pid,
        port,
        if force { " [FORCE]" } else { "" }
    );

    let killer = ProcessKiller::new();

    let result = if force {
        killer.kill(pid, true).await
    } else {
        killer.kill_gracefully(pid).await
    };

    match result {
        Ok(true) => {
            println!("✓ Process killed successfully.");
            Ok(())
        }
        Ok(false) => {
            println!("Process already terminated.");
            Ok(())
        }
        Err(e) => {
            bail!("Failed to kill process: {}", e);
        }
    }
}
